cassiopeia/solver/PercolationSolver.py

```diff
@@ -0,0 +1,350 @@
+        as a daughter clade of that node. This continues until each side of 
+        the partition is comprised only of single samples. If an algorithm 
+        cannot produce a split on a set of samples, then those samples are 
+        placed as sister nodes and the procedure terminates, generating a 
+        polytomy in the tree. This function will populate a tree inside the 
+        input CassiopeiaTree.
+
+        Args:
+            cassiopeia_tree: CassiopeiaTree storing a character matrix and
+                priors.
+        """
+
+        # A helper function that builds the subtree given a set of samples
+        def _solve(
+            samples: List[Union[str, int]],
+            tree: nx.DiGraph,
+            character_matrix: pd.DataFrame,
+            unique_character_matrix: pd.DataFrame,
+            priors: Dict[int, Dict[int, float]],
+            weights: Dict[int, Dict[int, float]],
+            missing_state_indicator: int,
+        ):
+            if len(samples) == 1:
+                return samples[0]
+            # Partitions the set of samples by percolating a similarity graph
+            clades = list(self.percolate(unique_character_matrix, samples, priors, weights, missing_state_indicator))
+            # Generates a root for this subtree with a unique int identifier
+            root = (
+                len(tree.nodes)
+                - unique_character_matrix.shape[0]
+                + character_matrix.shape[0]
+            )
+            tree.add_node(root)
+
+            for clade in clades:
+                if len(clade) == 0:
+                    clades.remove(clade)
+
+            # If unable to return a split, generate a polytomy and return
+            if len(clades) == 1:
+                for clade in clades[0]:
+                    tree.add_edge(root, clade)
+                return root
+            # Recursively generate the subtrees for each daughter clade
+            for clade in clades:
+                child = _solve(
+                    clade,
+                    tree,
+                    character_matrix,
+                    unique_character_matrix,
+                    priors,
+                    weights,
+                    missing_state_indicator,
+                )
+                tree.add_edge(root, child)
+            return root
+
+        weights = None
+        priors = None
+        if cassiopeia_tree.priors:
+            weights = solver_utilities.transform_priors(
+                cassiopeia_tree.priors, self.prior_transformation
+            )
+            priors = cassiopeia_tree.priors
+
+        # extract character matrix
+        character_matrix = cassiopeia_tree.get_original_character_matrix()
+        unique_character_matrix = character_matrix.drop_duplicates()
+
+        tree = nx.DiGraph()
+        samples = list(unique_character_matrix.index)
+        for i in samples:
+            tree.add_node(i)
```

Comment:
You can add all of these samples with one line of code: `tree.add_nodes_from(list(unique_character_matrix.index))`